tools/golangci-lint.sh

```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+# Check if golangci-lint is installed
+if ! command -v golangci-lint &> /dev/null; then
```

Comment:
Instead of having this script, there are predefined pre-commit hooks for golangci: https://github.com/golangci/golangci-lint/blob/master/.pre-commit-hooks.yaml

This has the advantage that it is separated from the rest of the installed programs on your laptop. And not everyone is using brew

Reply:
I have tried the provided hooks but there is a problem, if I don't pass the environment variables `GOOS=linux GOARCH=amd64` when triggering `golangci-lint` from my Mac it will identify that some of the assembly functions from 
`weaviate/adapters/repos/db/vector/hnsw/distancer/asm/dot_inline.go` and `weaviate/adapters/repos/db/vector/hnsw/distancer/asm/l2_inline.go` as unused:

```
git commit --amend
gofumpt..................................................................Passed
error groups.............................................................Passed
go routines..............................................................Passed
golangci-lint-full.......................................................Failed
- hook id: golangci-lint-full
- exit code: 1

adapters/repos/db/vector/hnsw/distancer/asm/dot_inline.go:33:6: func `dot3` is unused (unused)
func dot3[T, U number](x []T, y []T) U {
     ^
adapters/repos/db/vector/hnsw/distancer/asm/dot_inline.go:49:6: func `dot5` is unused (unused)
func dot5[T, U number](x []T, y []T) U {
     ^
adapters/repos/db/vector/hnsw/distancer/asm/dot_inline.go:63:6: func `dot7` is unused (unused)
func dot7[T, U number](x []T, y []T) U {
     ^
adapters/repos/db/vector/hnsw/distancer/asm/l2_inline.go:31:6: func `l23` is unused (unused)
func l23[T number, U number](x []T, y []T) U {
     ^
adapters/repos/db/vector/hnsw/distancer/asm/l2_inline.go:48:6: func `l25` is unused (unused)
func l25[T number, U number](x []T, y []T) U {
     ^

black................................................(no files to check)Skipped
```
The reason, according to ChatGPT is:
`Assembly Files and CGO: Assembly files, like those in distancer/asm/, often contain platform-specific code. If these assembly files are designed only for certain architectures or if CGO_ENABLED varies by environment, it can affect whether functions are flagged as unused.`

I investigated if there is any way to pass environment variables when running a pre-commit hook but it seems they get triggered as a child process and it doesn't allow passing any, as well as it won't be using the parent's process ones, so it will fail always. Relying on the external script is the only way I could come up to pass those environmetn variables.

Also, I have added couple of improvements in the script:

* The version used will be in sync with the one from the Github runner, this way we ensure that whatever appears in the CI is the same as from the pre-commit. And if brew isn't present in the system it will install using `go get`.

*  By default it will run the linter only on the modified files, not on the whole codebase (this way we ensure it will run fast), then in the CI it will get triggered for the whole golang codebase (in cases like this in which updating to a new version of `golangci-lint` uncovers several other linter errors.

Reply:
oki, thanks for the explanation!

Reply:
Maybe we can ask @asdine why those 5 functions are being considered by the linter as unused, but the rest of the functions in the file do not have that problem. Could it be the way the assembly code was built?

Reply:
@jfrancoa It's because these functions are generic helpers that can be used by all architectures but are only used by arm64 (`GOARCH=arm64`) for now.
It's a bit surprising that you are seeing this error on your mac though, (unless you have an intel CPU?), I would have expected an error on an amd64 architecture.